addons/stock/models/stock_quant.py

```diff
@@ -763,23 +745,25 @@ def _get_gather_domain(self, product_id, location_id, lot_id=None, package_id=No
             domain = expression.AND([['|', ('expiration_date', '>=', self.env.context['with_expiration']), ('expiration_date', '=', False)], domain])
         return domain
 
-    def _gather(self, product_id, location_id, lot_id=None, package_id=None, owner_id=None, strict=False, qty=0):
+    def _gather(self, product_id, location_id, lot_id=None, package_id=None, owner_id=None, strict=False, qty=0, quants_cache=None):
         """ if records in self, the records are filtered based on the wanted characteristics passed to this function
             if not, a search is done with all the characteristics passed.
         """
         removal_strategy = self._get_removal_strategy(product_id, location_id)
         domain = self._get_gather_domain(product_id, location_id, lot_id, package_id, owner_id, strict)
         domain, order = self._get_removal_strategy_domain_order(domain, removal_strategy, qty)
-        if self.ids:
-            sort_key = self._get_removal_strategy_sort_key(removal_strategy)
-            res = self.filtered_domain(domain).sorted(key=sort_key[0], reverse=sort_key[1])
+        if quants_cache is not None and strict:
+            res = quants_cache[product_id.id, location_id.id, lot_id and lot_id.id, package_id and package_id.id, owner_id and owner_id.id]
```

Comment:
TODO: quant_cache is not sorted base on removal_strategy. But we only call it for strict=True so I'm not sure it's important because the removal_strategy is use for reservation (strict=False)